File: android/src/main/kotlin/com/monorithm/monolens/MediaProbe.kt
package com.monorithm.monolens

import android.graphics.BitmapFactory
import android.media.MediaMetadataRetriever
import androidx.exifinterface.media.ExifInterface
import java.io.File

/**
 * Reads dimensions, duration and size off a file without decoding it.
 *
 * Both branches report *display* dimensions rather than stored ones: a still
 * carries an EXIF orientation tag and a clip carries a rotation hint, and in
 * both cases the stored buffer is often sideways relative to what the author
 * saw. Dart's `CropRect` resolves against these numbers, so the transformers
 * below have to work in the same upright space.
 */
internal object MediaProbe {

    fun probe(path: String): MediaInfo {
        val file = File(path)
        if (!file.exists()) throw MonolensException.notFound(path)
        val byteSize = file.length()

        return probeImage(path, byteSize)
            ?: probeVideo(path, byteSize)
            ?: throw MonolensException.unsupported(path)
    }

    private fun probeImage(path: String, byteSize: Long): MediaInfo? {
        val options = BitmapFactory.Options().apply { inJustDecodeBounds = true }
        BitmapFactory.decodeFile(path, options)
        if (options.outWidth <= 0 || options.outHeight <= 0) return null

        val quarterTurned = runCatching {
            when (ExifInterface(path).getAttributeInt(
                ExifInterface.TAG_ORIENTATION,
                ExifInterface.ORIENTATION_NORMAL,
            )) {
                ExifInterface.ORIENTATION_ROTATE_90,
                ExifInterface.ORIENTATION_ROTATE_270,
                ExifInterface.ORIENTATION_TRANSPOSE,
                ExifInterface.ORIENTATION_TRANSVERSE,
                -> true

                else -> false
            }
        }.getOrDefault(false)

        return MediaInfo(
            path = path,
            width = (if (quarterTurned) options.outHeight else options.outWidth).toLong(),
            height = (if (quarterTurned) options.outWidth else options.outHeight).toLong(),
            durationMs = null,
            byteSize = byteSize,
        )
    }

    private fun probeVideo(path: String, byteSize: Long): MediaInfo? {
        val retriever = MediaMetadataRetriever()
        return try {
            retriever.setDataSource(path)
            val width = retriever.int(MediaMetadataRetriever.METADATA_KEY_VIDEO_WIDTH)
            val height = retriever.int(MediaMetadataRetriever.METADATA_KEY_VIDEO_HEIGHT)
            if (width == null || height == null) return null

            // The rotation hint is metadata, not pixels: a portrait recording is
            // stored landscape with a 90 tag, and the player swaps it back.
            val rotation = retriever.int(MediaMetadataRetriever.METADATA_KEY_VIDEO_ROTATION) ?: 0
            val quarterTurned = rotation == 90 || rotation == 270

            MediaInfo(
                path = path,
                width = (if (quarterTurned) height else width).toLong(),
                height = (if (quarterTurned) width else height).toLong(),
                durationMs = retriever.int(MediaMetadataRetriever.METADATA_KEY_DURATION)?.toLong()
                    ?: 0L,
                byteSize = byteSize,
            )
        } catch (_: Exception) {
            null
        } finally {
            runCatching { retriever.release() }
        }
    }

    private fun MediaMetadataRetriever.int(key: Int): Int? =
        extractMetadata(key)?.toIntOrNull()
}

/**
 * The failures the host API reports back to Dart. The codes are what
 * `MediaEditException.code` carries, so Dart can branch on them — most
 * importantly `monolens/cancelled`, which the composer swallows rather than
 * surfacing as a failure.
 */
internal class MonolensException(
    val code: String,
    override val message: String,
) : Exception(message) {

    fun toFlutterError(): FlutterError = FlutterError(code, message, null)

    companion object {
        fun notFound(path: String) = MonolensException("monolens/not-found", "No file at $path")

        fun unsupported(path: String) =
            MonolensException("monolens/unsupported", "Unsupported media at $path")

        fun decodeFailed(message: String) =
            MonolensException("monolens/decode-failed", message)

        fun encodeFailed(message: String) =
            MonolensException("monolens/encode-failed", message)

        fun invalidRange(message: String) =
            MonolensException("monolens/invalid-range", message)

        fun exportFailed(message: String) =
            MonolensException("monolens/export-failed", message)

        fun cancelled() = MonolensException("monolens/cancelled", "Export cancelled")
    }
}

File: ios/monolens/Sources/monolens/MediaProbe.swift
import AVFoundation
import ImageIO
import UIKit

/// Reads dimensions, duration and size off a file without decoding it.
///
/// Both branches report *display* dimensions rather than stored ones: a still
/// carries an EXIF orientation and a clip carries a preferred transform, and in
/// both cases the stored buffer is often sideways relative to what the author
/// saw. Dart's `CropRect` resolves against these numbers, so the croppers below
/// have to work in the same upright space.
enum MediaProbe {

  static func probe(path: String) throws -> MediaInfo {
    let url = URL(fileURLWithPath: path)
    guard FileManager.default.fileExists(atPath: path) else {
      throw MonolensError.notFound(path)
    }
    let byteSize = fileSize(at: path)

    if let image = probeImage(url: url, byteSize: byteSize) {
      return image
    }
    if let video = probeVideo(url: url, byteSize: byteSize) {
      return video
    }
    throw MonolensError.unsupported(path)
  }

  private static func probeImage(url: URL, byteSize: Int64) -> MediaInfo? {
    guard let source = CGImageSourceCreateWithURL(url as CFURL, nil),
      CGImageSourceGetCount(source) > 0,
      let properties = CGImageSourceCopyPropertiesAtIndex(source, 0, nil) as? [CFString: Any],
      let width = properties[kCGImagePropertyPixelWidth] as? Int,
      let height = properties[kCGImagePropertyPixelHeight] as? Int
    else { return nil }

    // EXIF orientations 5–8 are the quarter-turned ones, where the stored
    // buffer's width is the displayed height.
    let orientation = properties[kCGImagePropertyOrientation] as? UInt32 ?? 1
    let turned = orientation >= 5 && orientation <= 8

    return MediaInfo(
      path: url.path,
      width: Int64(turned ? height : width),
      height: Int64(turned ? width : height),
      durationMs: nil,
      byteSize: byteSize
    )
  }

  private static func probeVideo(url: URL, byteSize: Int64) -> MediaInfo? {
    let asset = AVURLAsset(url: url)
    guard let track = asset.tracks(withMediaType: .video).first else { return nil }

    // naturalSize is the encoded size; preferredTransform is the rotation the
    // player applies. Multiplying gives what the author actually sees.
    let transformed = track.naturalSize.applying(track.preferredTransform)
    let width = abs(transformed.width)
    let height = abs(transformed.height)

    let seconds = CMTimeGetSeconds(asset.duration)
    let durationMs = seconds.isFinite ? Int64((seconds * 1000).rounded()) : 0

    return MediaInfo(
      path: url.path,
      width: Int64(width.rounded()),
      height: Int64(height.rounded()),
      durationMs: durationMs,
      byteSize: byteSize
    )
  }

  static func fileSize(at path: String) -> Int64 {
    let attributes = try? FileManager.default.attributesOfItem(atPath: path)
    return (attributes?[.size] as? NSNumber)?.int64Value ?? 0
  }
}

/// The failures the host API reports back to Dart. The codes are what
/// `MediaEditException.code` carries, so Dart can branch on them.
enum MonolensError: Error {
  case notFound(String)
  case unsupported(String)
  case decodeFailed(String)
  case encodeFailed(String)
  case invalidRange(String)
  case exportFailed(String)
  case cancelled

  var pigeonError: PigeonError {
    switch self {
    case .notFound(let path):
      return PigeonError(code: "monolens/not-found", message: "No file at \(path)", details: nil)
    case .unsupported(let path):
      return PigeonError(
        code: "monolens/unsupported", message: "Unsupported media at \(path)", details: nil)
    case .decodeFailed(let message):
      return PigeonError(code: "monolens/decode-failed", message: message, details: nil)
    case .encodeFailed(let message):
      return PigeonError(code: "monolens/encode-failed", message: message, details: nil)
    case .invalidRange(let message):
      return PigeonError(code: "monolens/invalid-range", message: message, details: nil)
    case .exportFailed(let message):
      return PigeonError(code: "monolens/export-failed", message: message, details: nil)
    case .cancelled:
      return PigeonError(code: "monolens/cancelled", message: "Export cancelled", details: nil)
    }
  }
}
